File: compiler/bench/lucas.c
// expected: 1388710083
// Lucas numbers - like Fibonacci but starts with 2, 1
#include <stdio.h>
#include <stdint.h>

long lucas(int n) {
    if (n == 0) return 2;
    if (n == 1) return 1;
    long a = 2, b = 1;
    for (int i = 1; i < n; i++) {
        long t = a + b;
        a = b;
        b = t;
    }
    return b;
}

int main(void) {
    long result = 0;
    for (int i = 0; i < 5000000; i++) {
        result ^= lucas(i % 45);
    }
    printf("%ld\n", result);
    return 0;
}

File: compiler/bench/lucas.fs
\ expected: 1388710083
\ Lucas numbers - like Fibonacci but starts with 2, 1

: lucas ( n -- L_n )
  dup 0= if drop 2 exit then
  dup 1 = if drop 1 exit then
  2 1                     \ a=L0=2, b=L1=1 (stack: a b)
  rot 1 do
    tuck +                \ (a b) -> (b a b) -> (b a+b)
  loop nip ;

: main
  0
  5000000 0 do
    i 45 mod lucas xor
  loop
  . cr ;
